internal/plugin/connectors/tcp/mssql/connector.go

```diff
@@ -199,56 +228,99 @@ func (connector *SingleUseConnector) Connect(
+	// driver on construction.
 	clientLoginChannel <- *clientLogin
 
-	// Block continuation until driver has completed connection
-	<- connectPhaseFinished
+	var loginAck mssql.LoginAckStruct
+	// Block continuation until error or driver has completed connection
+responseLoop:	for {
+		select {
+		// a protocol error is received from the server
+		case protocolErr := <- errChannel:
+			err = protocolErr
+			break responseLoop
+		// a loginAck is sent from the server, this will be followed by connectPhaseFinished
+		// unless there's an issue
+		case loginAck = <- loginAckChannel:
+			break
+		// connect phase is finished. this could be a non-protocol error or after loginAck
+		case res := <- connectPhaseFinished:
+			if res.err != nil {
+				err = res.err
+			} else {
+				// Verify the driverConn is an mssql driverConn object and get its
+				// underlying transport
+				connector.backendConn = res.conn.NetConn()
+			}
+
+			break responseLoop
+		}
+	}
+
 	if err != nil {
-		wrappedError := errors.Wrap(err, "failed to connect to mssql server")
-		connector.sendError(wrappedError)
-		return nil, wrappedError
+		connector.sendError(err)
+		return nil, err
 	}
 
-	// TODO: 	Send the login response to the client 	 (#1016)
-	// TODO: 	Verify appropriate errors are passed to the client (#1013)
-	if _, err = clientConn.Write(connector.CreateAuthenticationOKMessage()); err != nil {
+	if err = mssql.WriteLoginAck(connector.clientBuff, loginAck); err != nil {
 		wrappedError := errors.Wrap(
 			err,
 			"failed to send a successful authentication response to client",
 		)
 		connector.sendError(wrappedError)
 		return nil, wrappedError
 	}
-	// Verify the driverConn is an mssql driverConn object and get its underlying transport
-	mssqlConn := driverConn.(*mssql.Conn)
-	connector.backendConn = mssqlConn.NetConn()
+
+	// DONE: 	Send the login response to the client 	 (#1016)
+	// DONE: 	Verify appropriate errors are passed to the client (#1013)
 
 	return connector.backendConn, nil
 }
 
 
 // TODO: Add ability to receive an MSSQL error and send it to the client (#1013)
 func (connector *SingleUseConnector) sendError(err error) {
-	//NOTE: no need to log, Secretless already does this for every error sent back
-	//from the Connect method
-	//connector.logger.Error(err)
-	connector.sendErrorToClient()
+	var protocolErr mssql.Error
+	if _protocolErr, ok := err.(mssql.Error); ok {
+		protocolErr = _protocolErr
+	} else {
+		protocolErr = mssql.Error{
+			// SQL Error Number - currently using 18456 (login failed for user)
+			// TODO: Find generic error number
+			Number:     18456,
+			// State -
+			// TODO: better understand this.
+			State:      0x01,
+			// Severity Class - 16 indicates a general error that can be corrected by the user.
+			Class:      16,
+			Message:    errors.Wrap(err, "secretless").Error(),
+			ServerName: "secretless",
+			ProcName:   "",
+			LineNo:     0,
+		}
+	}
+	_ = mssql.WriteError72(connector.clientBuff, protocolErr)
```

Comment:
Great example where an inability to follow line of sight indicates a deeper problem.

There are two separate things happening here:
1. Creation of the `mssql.Error`
2. Sending the error to the client

If you create one func for 1. only (call it, say, `protocolError(err error) mssql.Error`, line of sight will naturally fall out of that.

Then this function will become simply:

```go
func (connector *SingleUseConnector) sendError(err error) {
	_ = mssql.WriteError72(connector.clientBuff, protocolError(err))
}
```
---
Totally separate issue: Can't  `mssql.WriteError72` fail?  And I understand that there's perhaps no point in trying to send _that_ failure to the client, but I'd think we want to at least log that we were unable to send error X to the client.